sanitycheck: "CMake Error: : System Error: File name too long"

As the title states, there are some tests that fail due to "CMake Error: : System Error: File name too long".

I'm using  eCryptfs on my Ubuntu 16.04 installation with the 4.10 kernel. There's a hard byte limit of 143 that it imposes since it uses the rest of the chars for encrypting.

> $ ~/src/zephyr/zephyr$ getconf NAME_MAX .
> 143
> $ ~/src/zephyr/zephyr$ getconf PATH_MAX .
> 4096

I'm still going through my sanitycheck, but I've found two so far: qemu_cortex_m3 and qemu_nios2. For example: 

> CMake Error: Cannot open file for write: /home/psidhu/src/zephyr/zephyr/sanity-out/qemu_cortex_m3/tests/application_development/gen_inc_file/test/CMakeFiles/_home_psidhu_src_zephyr_zephyr_sanity-out_qemu_cortex_m3_tests_application_development_gen_inc_file_test_zephyr_include_generated__file_bin_inc.dir/build.make.tmp
> CMake Error: : System Error: File name too long

As we can see, `_home_psidhu_src_zephyr_zephyr_sanity-out_qemu_cortex_m3_tests_application_development_gen_inc_file_test_zephyr_include_generated__file_bin_inc.dir` is 147 chars in this example.

--- a/cmake/extensions.cmake
+++ b/cmake/extensions.cmake
@@ -301,12 +301,17 @@ function(generate_inc_file_for_target
   # targets
 
   # But first create a unique name for the custom target
-  # Replace / with _ (driver/serial => driver_serial) and . with _
-  set(generated_target_name ${generated_file})
+  string(
+    RANDOM
+    LENGTH 8
+    random_chars
+    )
 
-  string(REPLACE "/" "_" generated_target_name ${generated_target_name})
-  string(REPLACE "." "_" generated_target_name ${generated_target_name})
-  string(REPLACE "@" "_" generated_target_name ${generated_target_name})
+  get_filename_component(basename ${generated_file} NAME)
+  string(REPLACE "." "_" basename ${basename})
+  string(REPLACE "@" "_" basename ${basename})
+
+  set(generated_target_name "gen_${basename}_${random_chars}")
 
   add_custom_target(${generated_target_name} DEPENDS ${generated_file})
   add_dependencies(${target} ${generated_target_name})

